source/common/router/upstream_request.cc

```diff
@@ -72,6 +244,28 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
   if (cluster_info.has_value()) {
     stream_info_.setUpstreamClusterInfo(*cluster_info);
   }
+  if (!allow_upstream_filters_) {
+    return;
+  }
+
+  // Set up the upstream filter manager.
+  filter_manager_callbacks_ = std::make_unique<UpstreamRequestFilterManagerCallbacks>(*this);
+  filter_manager_ = std::make_unique<UpstreamFilterManager>(
+      *filter_manager_callbacks_, parent_.callbacks()->dispatcher(),
+      parent_.callbacks()->connection(), parent_.callbacks()->streamId(),
+      parent_.callbacks()->account(), true, parent_.callbacks()->decoderBufferLimit(),
+      *parent_.cluster(), *this);
+  parent_.cluster()->createFilterChain(*filter_manager_);
+
+  // Similar to how the downstream filter chain ends with the router, the
+  // upstream filter chain ends with the encoder filter, which sends data to the
+  // codec. Create that manually here.
+  Http::FilterFactoryCb cb = [this](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    codec_filter_ = std::make_shared<CodecFilter>(*this);
```

Comment:
a question on why we didn't pass the `Http::FilterChainFactoryCallbacks` to the filter instead of UpstreamRequest object, I thought `UpstreamRequest` is Envoy internal implement, shouldn't expose to the filter.

Reply:
I think I get answer by my self https://github.com/envoyproxy/envoy/pull/22553/files/e755fee044c1ca990736a3ed605dfb729d6052d4#diff-889698b89a1023b7d44c1c6a715e0b05b7d92c399e0104998290ba4a605fadfdR73-R77, you plan to clean up this later :)

Reply:
so here I would like to move over to the HCM creating the filter chain, which will be shared code.  I've got a TODO for that already insofar as it can't be done until all the direct links between codec filter and upstream request go away.  Got 90% of that refactor done in the next follow-up :-)